wicker/core/datasets.py

```diff
@@ -22,6 +24,154 @@
+        to fetch size in bytes on s3. Tuple index 0 is bucket and index 1 is key of the file.
+
+    Returns:
+        int size of file list in bytes.
+    """
+    # declare the variables neccessary to split
+    # out between multiple procs
+    manager = Manager()
+    total_size = manager.Value("i", 0)
+    lock = manager.Lock()
+    buckets_keys_list = list(buckets_keys)
+
+    buckets_keys_chunks = chunk_data_for_split(
+        chunkable_data=buckets_keys_list, chunk_additions=[total_size, lock], chunk_number=200
+    )
+
+    print("Grabbing file information from s3 heads")
+    # set the proc pool up and distribute the data between
+    # along with shared values
+    # reserve one cpu for other operations
+    pool = Pool(cpu_count() - 1)
+    # arbitrary chunking breaks typing here, still keep typing on function
+    # for documentation
+    pool.map(get_file_size_s3_threaded, buckets_keys_chunks)  # type: ignore
+    pool.close()
+    pool.join()
+    return total_size.value
+
+
+def get_file_size_s3_threaded(input_tuple: Tuple[List[Tuple[str, str]], ValueProxy, Lock]):
+    """Get file size of a list of s3 paths.
+
+    This function takes the input in the form of a tuple for efficient passing in a mutli
+    processing setting. The input buckets_keys_chunks_local are a list of tuples that is split
+    out across multiple threads to reduce the i/o wait to minimum on one single processor.
+    The general use case is for saturating the i/o from s3 in order to most efficiently caclulate
+    the size of the file tuple list in bytes.
+
+    Tuple structure - The tuple contains 3 parts denoted below.
+
+        buckets_keys_chunks_local - The list of tuples denoting bucket and key of files on s3 to
+        parse. Generally column files but will work with any data.
+
+        sum_value - A python shared memory object (ctype) for storing the byte sum across all processes.
+
+        lock - A shared memory lock for locking the sum_value during write procedure.
+
+    Args:
+        input_tuple: A tuple containing the relevant pieces of data used to grab and store byte size.
+            Works across procs so needs a lock and shared memory space.
+
+    Returns:
+        Void function - does all modifications in place on the shared memory value.
+
+    """
+    buckets_keys_chunks_local, sum_value, lock = input_tuple
+
+    # chunk the process again between multiple threads on each proc
+    # done to reduce i/o wait on each individual proc
+    local_chunks = chunk_data_for_split(chunkable_data=buckets_keys_chunks_local, chunk_number=200)
+    # set up thread pool with max threads
+    thread_pool = ThreadPool()
+
+    # run the threads and sum the results together
+    # same as with multi proc, arbitrary chunking breaks typing, keep for docs
+    result = sum(list(tqdm.tqdm(thread_pool.map(iterate_bucket_key_chunk_for_size, local_chunks))))  # type: ignore
+    with lock:
```

Comment:
So now you won't need the lock, you will just return `result` up to the `pool.map` that called `get_file_size_s3_threaded`.

Reply:
Yep, removed.